vesc_hw_interface/src/vesc_step_difference.cpp

```diff
@@ -0,0 +1,141 @@
+/*********************************************************************
+ * Copyright (c) 2023 SoftBank Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ********************************************************************/
+
+#include "vesc_hw_interface/vesc_step_difference.h"
+namespace vesc_step_difference
+{
+/**
+ * Hall sensors attached to brushless motors have low angular resolution.
+ * When looking at the time variation of the counter value obtained from the Hall sensor,
+ * the value is slow and oscillatory. (The value will be either 0 or 1)
+ * To prevent this, when the speed is slow, the sampling time is increased and the time variation of the counter value
+ * is smoothed. This means that the smoothing filter is applied only at low speeds.
+ */
+VescStepDifference::VescStepDifference()
+{
+}
+
+VescStepDifference::~VescStepDifference()
+{
+}
+
+/**
+ * Enables smoothing and initializes internal parameters.
+ * If this function is not called, smoothing is disabled and the getStepDifference function outputs the difference from
+ * the previous step as is.
+ */
+void VescStepDifference::enableSmooth(double control_rate, double max_sampling_time, int max_step_diff)
+{
+  step_diff_vw_max_step_ = max_step_diff;
+  int step_diff_vw_max_window_size_ = static_cast<int>(std::round(control_rate * max_sampling_time));
+  if (step_diff_vw_max_window_size_ > 0)
+  {
+    enable_smooth_ = true;
+    step_input_queue_.resize(step_diff_vw_max_window_size_ + 1);
+  }
+  else
+  {
+    // Disable smoothing if step_diff_vw_max_window_size_ is too small
```

Comment:
It would be great if you could print a log saying the `enable_smooth` variable is set to false due to a small window size or something.